Adaptive policies research

ohadeytan opened the pull request:
Adaptive policies for our upcoming Middleware2018 article: "Adaptive Software Cache Management".
More details will be provided [here](https://www.cs.technion.ac.il/~ohadey/adaptive-caching/AdaptiveSoftwareCacheManagementMiddleware2018.html).

ben-manes commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/sketch/climbing/IndicatorClimber.java:
```diff
@@ -0,0 +1,55 @@
+      }
+      return new Adaptation(Adaptation.Type.DECREASE_WINDOW, (int) ((oldPercent - newPercent) * cacheSize));
+    }
+    return new Adaptation(Adaptation.Type.HOLD, 0);
```
`return Adaptation.HOLD` for the constant (I know this api sucks)

ben-manes commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/sketch/climbing/MiniSimClimber.java:
```diff
@@ -0,0 +1,88 @@
+  public void doAlways(long key) {
+    sample++;
+    String skey = String.valueOf(key);
+    if (Murmur3.hash_x86_32(skey.getBytes(), skey.length(), 0x7f3a2142) % R < 1) {
```
Guava case described above

ben-manes commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/sketch/climbing/MiniSimClimber.java:
```diff
@@ -0,0 +1,88 @@
+
+  @Override
+  public Adaptation adapt(int windowSize, int protectedSize) {
+    if (sample > 1000000) {
```
This means it will try to adapt every access after `1000000`? Should it be a modulus? Should it be a configuration constant?

ben-manes commented:
@Maaartinus fyi since you enjoy these types of topics 😄

ohadeytan commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/sketch/Indicator.java:
```diff
@@ -0,0 +1,136 @@
+    StreamSummary<Long> stream;
+
+    public EstSkew() {
+      this.stream = new StreamSummary<>(1000);
```
Currently, the idea is that the Indicator is trace dependent rather than cache size dependent. But this could be debatable. For now, I leave it as is.

ben-manes replied:
Can we pull it into the config file?

ohadeytan commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/irr/HillClimberFrdPolicy.java:
```diff
@@ -0,0 +1,424 @@
+
+  private void onFilterHit(Node node) {
+    // Cache hit in the filter stack: Move the corresponding block to the MRU
+    // position of the filter
```
I just followed your FRD format. I'll fix it.

ohadeytan commented:
Few constants here could be configurable, but I didn't want to add too many options. Do you think we should allow as much as possible configurations? 

Regarding the separation vs. abstraction of policies, I'm not sure what is the best way. IMO separation makes it simpler and clearer when you don't want to change configuration options, but when you start to go deeper and change things it is more reasonable to make an abstraction and merge policies. 

For now, I'll merge it as is and perhaps we would reorganize few things in the future.

ben-manes commented:
@ohadeytan I cleaned things up a bit. I don't know enough about are the constants except that they look a bit scary.

For `MiniSimClimber`, should that be a modulus? Or is it a warm-up period and you expect it to adapt on every operation? It's not obvious from the code.

For 'Indicator', what would a good name be for the stream constant?

ohadeytan commented:
I extracted all the constants to the configuration file and fixed some styling and a bug I have found in minisim.
Hope it's clearer now.
Sorry for the delay.

ohadeytan commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/irr/HillClimberFrdPolicy.java:
```diff
@@ -0,0 +1,424 @@
+    }
+
+    public double percentMain() {
+      return config().getDouble("frd.percent-main");
```
Since all of them are frd derivations I think it makes sense to leave it as is.

ohadeytan commented on simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/irr/IndicatorFrdPolicy.java:
```diff
@@ -0,0 +1,383 @@
+  
+  private void adapt(long key) {
+	indicator.record(key);
+	if (indicator.getSample() == 50000) {
```
Done.

ben-manes commented:
Thanks! 😄
